### ISSUE TITLE
window_events only appears to emit first event

### ISSUE THREAD

**Issue by jgrund** [seed-rs/seed#138]
Consider:

```rust
        fn window_events(model: &Model) -> Vec<seed::dom_types::Listener<Msg>> {
            vec![
                mouse_ev(Ev::Click, |_| Msg::Increment),
                mouse_ev(Ev::Click, |_| Msg::Increment),
                mouse_ev(Ev::Click, |_| Msg::Increment),
                mouse_ev(Ev::Click, |_| Msg::Increment),
                mouse_ev(Ev::Click, |_| Msg::Increment),
            ]
        }
```

Where `Msg::Increment` just increments a counter on the model that starts at 0.

When a click event occurs, I'd expect the counter to increment to 5. Instead it only increments to 1.

Interestingly, all 5 listeners can be seen on the window:

<img width="792" alt="Screen Shot 2019-06-13 at 2 54 57 PM" src="https://user-images.githubusercontent.com/458717/59459641-479fd980-8deb-11e9-809a-974869badded.png">

I think there may be some sort of listener detach / re-attach that occurs after the first event and somehow causes the rest to be skipped.

**Comment by jgrund** [comment]
After the first event triggers an `update`, `setup_window_listeners` is called, which in turn detaches / reattaches the events.

Once the detach occurs, the remaining events are no longer registered as listeners and are skipped.

Commenting out `setup_window_listeners` here:

https://github.com/jgrund/seed/blob/48b814541242aab8166d0801fef21b7dbd1b5a32/src/vdom.rs#L449

Allows the counter to increment to 5 as expected.

**Comment by David-OConnor** [comment]
I can reproduce. Removing that line isn't the solution, because it's required for the listeners to update when the model changes.

**Comment by jgrund** [comment]
Could extra info be added regarding [`currentTarget`](https://developer.mozilla.org/en-US/docs/Web/API/Event/currentTarget) being window, and skip detach / reattach if that's the case?

I don't have enough context to know if that will have follow-on effects if skipping under certain situations...

**Comment by David-OConnor** [comment]
Could try that. Need to verify this doesn't also affect normal elements that have multiple listeners. If it does, may need to internally lump them into one

**Comment by jgrund** [comment]
Temporary workaround is to delegate the handling to the update fn by registering a single handler per event type needed. E.g.:

```
fn window_events(_model: &Model) -> Vec<seed::dom_types::Listener<Msg>> {
    vec![simple_ev(Ev::Click, Msg::WindowClick)]
}
```

**Comment by David-OConnor** [comment]
Concur on your workaround until fixed. As you observed, the process is being interrupted between update calls. Appears to only affect window events, where there are multiple listeners for the same trigger.

**Comment by MartinKavik** [comment]
If I understand correctly, the main problem is that we can't diff event listeners?

I was looking for some inspiration in [dodrio](https://github.com/fitzgen/dodrio) and on the first look:
- [diff algorithm](https://github.com/fitzgen/dodrio/blob/master/src/diff.rs#L160) - it compares event ids and if both listeners handle the same event, then it only replaces old callback with the new one
- [update_event_listener](https://github.com/fitzgen/dodrio/blob/master/src/change_list/mod.rs#L310)
- [struct Listener](https://github.com/fitzgen/dodrio/blob/master/src/node.rs#L122)

I don't know current Seed implementation and how much abstraction is hiding between those VirtualDOM listeners and native DOM listeners, but I think it should help with this issue and probably improve performance.

**Comment by David-OConnor** [comment]
I'll take a look. Of note, this doesn't appear to affect multiple same-trigger listeners on normal elements.

**Comment by MartinKavik** [comment]
> Of note, this bug doesn't appear to affect multiple same-trigger listeners on normal elements.
My first guess: Maybe because we reattach window listeners in every update but diff of normal elements can be skipped (whole render is skipped or some of parents?).
Good luck, I'm working on the realworld example now, if you want some help / code review, etc. just ping me.

**Comment by David-OConnor** [comment]
I'm leaning towards a breaking change here: `Msg` must implement `PartialEq`, and the event constructors other than `raw_ev` will have signature `(trigger, message)`, ie what `simple_ev` has now. This will allow us to diff most listeners completely. In the event more complex behavior is desired in the view, you can still use `raw_ev`, and accept the slower diffing. (And possible errors like this until resolved).

What do you think? I have most of it implemented already.

**Comment by MartinKavik** [comment]
Why you want to add `PartialEq` for `Msg`? Trigger/event isn't enough? I think it would be annoying - you would have to write custom implementation `PartialEq` for yoour `Msg`s because you can't derive it automatically in many cases.

**Comment by MartinKavik** [comment]
I want to know how do you pass data from event (e.g. changed input text) to message / update function / model when the new signature is `(trigger, message)`?

Look at [my rust playground example](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=8f05c6ee602a5fd39181c92b64992dd0) or at the [documentation](https://doc.rust-lang.org/std/mem/fn.discriminant.html) - it should resolve _"`PartialEq` problem"_ if we assume that `Msg` is always `enum`.

**Comment by David-OConnor** [comment]
Nice workaround with mem; (Related: the `dodrio` implementation you linked earlier appeared to use mem to diff the closures, but I didn't understand it) I hadn't considered cases where other data types are included in the `Msg` variants, which would indeed be annoying.

I'll post the branch once I've got it working, or if unable to fix the obstacle I'm stuck on.

**Comment by MartinKavik** [comment]
> Related: the dodrio implementation you linked earlier appeared to use mem to diff the closures, but I didn't understand it

Could you send link to the code please?

**Comment by David-OConnor** [comment]
https://github.com/fitzgen/dodrio/blob/master/src/node.rs#L254

**Comment by MartinKavik** [comment]
I think it doesn't do diff but only checks that we can do magical destructuring through `union`.
So method `get_callback_parts` basically converts `Listener` to `(u32, u32)` (I don't know what exactly  these numbers mean but I think that the first one can be a pointer to the listener's callback (which is 0 when the function is dropped(?)).
"callback parts" are also used as a key for a hashmap, where the values are callbacks - it seems that this hashmap is used as the link between Js and Rust world, so we can change callbacks in the Rust world however they can be invoked from the JS world by those "callback parts".
You can experiment with destructuring here: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=06f0883cc21579f9549691ec7a28f969

**Comment by MartinKavik** [comment]
I was trying to find some solutions for this issue and this is the result:

- It's impossible to reliable compare handlers (aka functions or closures). We can compare functions by reference and closures (`Fn*` traits) by reference in vtable in some cases, but we cannot be sure that they are identical and the side effects would be the same. Other ways are to serialize and hash them or add arbitrary key/hash into listeners but these are very cumbersome solutions. 

- How the most of other frameworks (virtual DOMs) resolve problems with event listeners:
   - **A)** They have one event listener per event (aka trigger) in a DOM node. This one event listener invokes registered user handlers with the same event in the given node.
   - **B)** They have one event listener per event in a document/window/root node. All events are handled by those root listeners.

- I think the first option **(A)** is more suitable to Seed architecture and simpler. 
As a result, VirtualDOM should be faster because we won't need to reattach all listeners in each patching and it'll solve the original issue _"window_event only appears to emit first event"_ because we won't be killing listeners while they are firing events.

So.. I want to take it as a long-term homework and refactor Seed's VirtualDOM a little bit during the process (delete hooks and deprecated event trigger (a part of the old `seed::update`), allow nodes like `comment` to be included in `raw!`, etc.).
@jgrund, @David-OConnor - do you see any problems with this solution?

**Comment by MartinKavik** [comment]
Just a note: There is an old idea related to VirtualDOM - https://github.com/David-OConnor/seed/issues/180.

**Comment by MartinKavik** [comment]
@jgrund Fix should be in `master` - could you confirm that it works for you and eventually close the issue? Thanks.
(PR https://github.com/seed-rs/seed/pull/330)

**Comment by MartinKavik** [comment]
Should be fixed in the latest release (0.6.0). 
We can reopen the issue when some problems appear.
